plugin/db/util/driverutil.go

```diff
@@ -216,36 +205,41 @@ func beginMigration(ctx context.Context, executor MigrationExecutor, m *db.Migra
 		}
 	}
 
-	// Check if there is any higher version already been applied
-	if version, err := executor.CheckOutOfOrderVersion(ctx, tx, m.Namespace, m.Source, m.Version); err != nil {
+	sqldb, err := executor.GetDbConnection(ctx, bytebaseDatabase)
+	if err != nil {
 		return -1, err
-	} else if version != nil && len(*version) > 0 {
-		// Clickhouse will always return non-nil version with empty string.
-		return -1, common.Errorf(common.MigrationOutOfOrder, fmt.Errorf("database %q has already applied version %s which is higher than %s", m.Database, *version, m.Version))
 	}
-
-	// If the migration engine is VCS and type is not baseline and is not branch, then we can only proceed if there is existing baseline
-	// This check is also wrapped in transaction to avoid edge case where two baselinings are running concurrently.
-	if m.Source == db.VCS && m.Type != db.Baseline && m.Type != db.Branch {
-		if hasBaseline, err := executor.FindBaseline(ctx, tx, m.Namespace); err != nil {
-			return -1, err
-		} else if !hasBaseline {
-			return -1, common.Errorf(common.MigrationBaselineMissing, fmt.Errorf("%s has not created migration baseline yet", m.Database))
-		}
+	// From concurrency perspective, there's no difference of using transaction or not. However, we use transaction here to save some code of starting a transaction.
+	tx, err := sqldb.BeginTx(ctx, nil)
+	if err != nil {
+		return -1, err
 	}
+	defer tx.Rollback()
 
-	// VCS based SQL migration requires existing baselining
-	requireBaseline := m.Source == db.VCS && m.Type == db.Migrate
-	sequence, err := executor.FindNextSequence(ctx, tx, m.Namespace, requireBaseline)
+	largestSequence, err := executor.FindLargestSequence(ctx, tx, m.Namespace, false /* baseline */)
+	if err != nil {
+		return -1, err
+	}
+	largestBaselineSequence, err := executor.FindLargestSequence(ctx, tx, m.Namespace, true /* baseline */)
```

Comment:
I feel we only need FindLargestSequenceSinceLastBaseline and don't need the extra baseline parameter.

Then we call FindLargestVersionSinceLastBaseline and we also don't need to pass "largestBaselineSequence".

e.g.

sequence, version
1, 1.1.0
2, 1.1.3
3,  1.1.2

FindLargestSequenceSinceLastBaseline returns 3 and FindLargestVersionSinceLastBaseline returns 1.1.3
 
FindLargestVersionSinceLastBaseline doesn't depend on the returned sequence number 3 to get 1.1.3

We need to do a 2 step query for each method.

1. Find the id for the last baseline.
2. Search the migration record after that id (or if there is no baseline, search from the beginning)

Reply:
We would need to find the sequence of the last baseline/branch. And we may prefer to use sequence than id to leverage the index. Also, this implementation saves additional interface methods on the executor.

Reply:
>> We would need to find the sequence of the last baseline/branch. And we may prefer to use sequence than id to leverage the index

Using sequence is better.

My question is in GetLargestVersionSinceLastBaseline

`
		SELECT MAX(version) FROM bytebase.public.migration_history
		WHERE namespace = ? AND sequence >= ?
	`

The version we get may NOT be the largest version since the last baseline. In the above example

1, 1.1.0
2, 1.1.3
3, 1.1.2

This method would return version 1.1.2 for the highest sequence 3, while the actual highest version since the last baseline is 1.1.3.


>> Also, this implementation saves additional interface methods on the executor.

It's still 2 methods, and we can get rid of the boolean baseline flag to FindLargestSequence since GetLargestVersionSinceLastBaseline does not need to use largestBaselineSequence. In another word, the finding of the largest sequence is independent of finding the largest version since the last baseline.

Reply:
Done.